runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -10341,6 +10341,473 @@ inlineCompareAndSwapNative(
+
+      // Otherwise, load the byte at address [buf + (limit - 1)] (which is the same byte again in the 1 byte case)
+      generateRegMemInstruction(TR::InstOpCode::L1RegMem, node, chunkReg, generateX86MemoryReference(bufReg, limitReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes() - 1, cg), cg);
+
+      // If the byte is positive, jump immediately to returnNoNegativesLabel
+      generateRegImmInstruction(TR::InstOpCode::CMP1RegImm1, node, chunkReg, 0, cg);
+      generateLabelInstruction(TR::InstOpCode::JG4, node, returnNoNegativesLabel, cg);
+
+      // Otherwise, set the index to limit - 1 and jump to returnHasNegativesLabel
+      // generateRegMemInstruction(TR::InstOpCode::LEA4RegMem, node, indexReg, generateX86MemoryReference(limitReg, -1, cg), cg);
+      generateRegInstruction(TR::InstOpCode::INC4Reg, node, indexReg, cg);
+      generateLabelInstruction(TR::InstOpCode::JMP4, node, returnHasNegativesLabel, cg);
+      }
+
+
+   // Case in which there are three or more residual bytes
+   generateLabelInstruction(TR::InstOpCode::label, node, threeOrMoreBytesLabel, cg);
+
+   // if loopLimit > 4, jump to fiveOrMoreBytesLabel
+   generateRegImmInstruction(TR::InstOpCode::CMP4RegImm4, node, loopLimitReg, 4, cg);
+   generateLabelInstruction(TR::InstOpCode::JG4, node, fiveOrMoreBytesLabel, cg);
+
+   if (recognizedMethod == TR::java_lang_StringCoding_hasNegatives)
+      {
+      // Load the first two bytes at address [buf + index] into the chunk register
+      generateRegMemInstruction(TR::InstOpCode::L2RegMem, node, chunkReg, generateX86MemoryReference(bufReg, indexReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg), cg);
+      // OR the second two bytes at address [buf + (limit - 2)] into the chunk register
+      generateRegMemInstruction(TR::InstOpCode::OR2RegMem, node, chunkReg, generateX86MemoryReference(bufReg, limitReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes() - 2, cg), cg);
+      generateLabelInstruction(TR::InstOpCode::JMP4, node, residualTestLabel, cg);
+      }
+   else
+      {
+      // Load the first two bytes at address [buf + index] into the chunk register
+      generateRegMemInstruction(TR::InstOpCode::L2RegMem, node, chunkReg, generateX86MemoryReference(bufReg, indexReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg), cg);
+
+      // AND the chunk with the mask
+      generateRegRegInstruction(TR::InstOpCode::AND2RegReg, node, chunkReg, maskReg, cg);
+      // If the result is zero, there are no negatives in these first two bytes, so jump to the second load
+      generateRegRegInstruction(TR::InstOpCode::TEST2RegReg, node, chunkReg, chunkReg, cg);
+      generateLabelInstruction(TR::InstOpCode::JE4, node, threeOrMoreSecondLoadLabel, cg);
+
+      // Otherwise, count the number of zeroes in the chunk before the first sign bit
+      generateRegRegInstruction(TR::InstOpCode::TZCNT2RegReg, node, chunkReg, chunkReg, cg);
```

Comment:
Should this be using `lzcnt` instead?  I think the main loop that uses `pmovmskb` needs to use `tzcnt` operations because of the order in which the bits are placed in the target register.  In this case, I think you want to know the position of the first sign bit, whereas `tznct` will give you the position of the last sign bit.

Reply:
Since x86 is little endian (as far as I can tell that's something that we can always rely on), `tzcnt` does actually find the first sign bit, I've tested it out to make sure.

Reply:
Oops!  I hadn't been thinking about the fact that those bytes would be reversed in loading them into a register.